internal/controller/linodefirewall_controller_helpers.go

```diff
@@ -382,44 +384,56 @@ func processIPv4Rules(ips []string, rule infrav1alpha2.FirewallRule, ruleLabel s
 		return
 	}
 
-	ipv4chunks := chunkIPs(ips)
-	for i, chunk := range ipv4chunks {
-		v4chunk := chunk
-		*rules = append(*rules, linodego.FirewallRule{
-			Action:      rule.Action,
-			Label:       ruleLabel,
-			Description: fmt.Sprintf("Rule %d, Created by CAPL: %s", i, rule.Label),
-			Protocol:    rule.Protocol,
-			Ports:       rule.Ports,
-			Addresses:   linodego.NetworkAddresses{IPv4: &v4chunk},
-		})
+	ipchunks := chunkIPs(ips)
+	if ipType == linodego.IPTypeIPv4 {
+		for i, chunk := range ipchunks {
+			*rules = append(*rules, linodego.FirewallRule{
+				Action:      rule.Action,
+				Label:       ruleLabel,
+				Description: fmt.Sprintf("Rule %d, Created by CAPL: %s", i, rule.Label),
+				Protocol:    rule.Protocol,
+				Ports:       rule.Ports,
+				Addresses:   linodego.NetworkAddresses{IPv4: &chunk},
+			})
+		}
+	} else if ipType == linodego.IPTypeIPv6 {
+		for i, chunk := range ipchunks {
+			*rules = append(*rules, linodego.FirewallRule{
+				Action:      rule.Action,
+				Label:       ruleLabel,
+				Description: fmt.Sprintf("Rule %d, Created by CAPL: %s", i, rule.Label),
+				Protocol:    rule.Protocol,
+				Ports:       rule.Ports,
+				Addresses:   linodego.NetworkAddresses{IPv6: &chunk},
+			})
+		}
 	}
 }
 
-// processIPv6Rules processes IPv6 rules and adds them to the rules slice
-func processIPv6Rules(ips []string, rule infrav1alpha2.FirewallRule, ruleLabel string, rules *[]linodego.FirewallRule) {
-	// Initialize rules if nil
-	if *rules == nil {
-		*rules = make([]linodego.FirewallRule, 0)
+func processFirewallRule(ctx context.Context, k8sClient clients.K8sClient, log logr.Logger, ruleRef *corev1.ObjectReference, firewall *infrav1alpha2.LinodeFirewall, ruleType string, createOpts *linodego.FirewallCreateOptions) error {
+	rule := &infrav1alpha2.FirewallRule{}
+	if ruleRef.Namespace == "" {
+		ruleRef.Namespace = firewall.Namespace
 	}
+	if k8sClient != nil {
+		if err := k8sClient.Get(ctx, client.ObjectKey{Namespace: ruleRef.Namespace, Name: ruleRef.Name}, rule); err != nil {
+			log.Error(err, "failed to fetch referenced AddressSet", "namespace", ruleRef.Namespace, "name", ruleRef.Name)
 
-	// If no IPs, return early
-	if len(ips) == 0 {
-		return
+			return err
+		}
+		finalizer := getFinalizer(firewall)
+		if !controllerutil.ContainsFinalizer(rule, finalizer) {
+			controllerutil.AddFinalizer(rule, finalizer)
+			if err := k8sClient.Update(ctx, rule); err != nil {
```

Comment:
Same as [here](TODO), but for adding the `LinodeFirewall` finalizer on the `FirewallRule` when creating multiple `LinodeFirewall`s referencing the same `FirewallRule`